test/spanner.ts

```diff
@@ -3260,6 +3261,48 @@ describe('Spanner with mock server', () => {
       assert.strictEqual(commitRequest.mutations.length, 1);
     });
 
+    it('should apply blind writes only once with mutations', async () => {
+      const database = newTestDatabase();
+      const mutations = new MutationSet();
+      mutations.upsert('Singers', {
+        SingerId: 1,
+        FirstName: 'Scarlet',
+        LastName: 'Terry',
+      });
+      mutations.upsert('Singers', {
+        SingerId: 2,
+        FirstName: 'Marc',
+      });
+      await database.writeAtLeastOnce(mutations, {});
+      await database.close();
+
+      // Verify that we don't have a BeginTransaction request for the retry.
+      const beginTxnRequest = spannerMock.getRequests().find(val => {
+        return (val as v1.BeginTransactionRequest).options?.readWrite;
+      }) as v1.BeginTransactionRequest;
+      assert.deepStrictEqual(beginTxnRequest, undefined);
+
+      // Verify that we have a single Commit request, and that the Commit request contains only one mutation.
+      assert.strictEqual(
+        1,
+        spannerMock.getRequests().filter(val => {
+          return (val as v1.CommitRequest).mutations;
+        }).length
+      );
+      const commitRequest = spannerMock.getRequests().find(val => {
+        return (val as v1.CommitRequest).mutations;
+      }) as v1.CommitRequest;
+      assert.ok(commitRequest, 'Commit was called');
+      assert.strictEqual(commitRequest.mutations.length, 2);
+      const request = spannerMock.getRequests().find(val => {
+        return (val as v1.CommitRequest).singleUseTransaction?.readWrite;
+      }) as v1.CommitRequest;
```

Comment:
You can combine the two `spannerMock.getRequests()` into one and filter for __both__ `mutations` and `singleUseTransaction?.readWrite`. That simplifies the code a bit, but it also ensures that there is really just one commit request. (In theory, the current verification also covers the strange case where we have two commit requests; one that contains the mutations, but without a single-use read/write transaction, and one that contains no mutations, but has a single-use read/write transaction.)

Reply:
Updated.